libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -0,0 +1,2036 @@
+
+    /* If we failed anywhere in the connection process, then destroy the semaphores created. */
+    if(status != IOT_HTTPS_OK)
+    {
+        /* If there was a connect was successful, disconnect from the network.  */
+        if(( _httpsConnection != NULL) && (_httpsConnection->isConnected))
+        {   
+            if (_networkDisconnect(_httpsConnection) != IOT_HTTPS_OK)
+            {
+                IotLogError("Failed to disconnect from a failed connection context initialization.");
+            }
+        }
+
+        if(connSemCreated)
+        {
+            IotSemaphore_Destroy( &(_httpsConnection->connSem) );
+        }
+
+        if(rxStartSemCreated)
+        {
+            IotSemaphore_Destroy( &(_httpsConnection->rxStartSem) );
+        }
+
+        if(rxFinishedSemCreated)
+        {
+            IotSemaphore_Destroy( &(_httpsConnection->rxFinishSem) );
+        }
+
+        /* Set the connection handle as NULL if everything failed. */
+        *pConnHandle = NULL;
+    }
+
+    IOT_FUNCTION_CLEANUP_END();
+}
+
+/* --------------------------------------------------------- */
+
+IotHttpsReturnCode_t IotHttpsClient_Connect(IotHttpsConnectionHandle_t * pConnHandle, IotHttpsConnectionInfo_t *pConnInfo)
+{
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    /* Check for NULL parameters in a public API. */
+    if((pConnHandle == NULL) || (pConnInfo == NULL))
+    {
+        IotLogError("Null parameters passed into IotHttpsClient_Connect");
+        status = IOT_HTTPS_INVALID_PARAMETER;
+    }
+
+    /* If a valid connection handle is passed in. */
+    if((status == IOT_HTTPS_OK) && (*pConnHandle != NULL))
+    {
+        /* If the handle in a connected state, then we want to disconnect before reconnecting. The ONLY way to put the 
+           handle is a disconnect state is to call IotHttpsClient_Disconnect(). */
+        if((*pConnHandle)->isConnected)
+        {
+            status = IotHttpsClient_Disconnect(*pConnHandle);
+            if(status != IOT_HTTPS_OK)
+            {
+                IotLogError("Error disconnecting a connected *pConnHandle passed to IotHttpsClient_Connect().Error code %d", status);
+            }
+        }
+    }
+
+    /* Connect to the server now. Initialize all resources needed for the connection context as well here. */
+    if(status == IOT_HTTPS_OK)
+    {
+        status = _connectHttpsServer(pConnHandle, pConnInfo);
+        if(status != IOT_HTTPS_OK)
+        {
+            IotLogError("Error in IotHttpsClient_Connect(). Error code %d.", status);
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static IotHttpsReturnCode_t _networkDisconnect(_httpsConnection_t* _httpsConnection)
+{
+    IotNetworkError_t networkStatus = IOT_NETWORK_SUCCESS;
+    IotHttpsReturnCode_t status = IOT_HTTPS_OK;
+
+    networkStatus = _httpsConnection->pNetworkInterface->close( _httpsConnection->pNetworkConnection );
```

Comment:
Is calling close() sufficient ? close() calls shutdown, but actually does not close the socket. When close is called, _FLAG_SHUTDOWN is set by network abstraction code, but destroyConnection is not set in the receive thread. It only makes sure that the receive task has exits. I think you need to call destroy explicitly to close the socket.

Reply:
There is no destroy operation. This does actually close the connection. It is all that is done in MQTT:
https://github.com/aws/amazon-freertos/blob/master/libraries/c_sdk/standard/mqtt/src/iot_mqtt_network.c#L823

Reply:
There is destroy operation:
https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/platform/freertos/iot_network_freertos.c#L597